<h2>Nextflow pipeline for whole genome sequence analysis of Human genome (Chr2) using NanoPore/Illumina Hybrid sequence </h2>

<h3>Software used in the pipeline.</h3>

**FastQC** is widely used and is robust, efficient, and versatile quality control software for a varied range of raw genetic data. It outputs a quality report which can be viewed to give an indication on how good the respective reads are. For more information, please visit the FastQC website https://www.bioinformatics.babraham.ac.uk/projects/fastqc/.

**NanoPlot** is a plotting tool for long read sequencing data and alignments. https://github.com/wdecoster/NanoPlot

**Filtlong** is a tool for filtering long reads by quality. It can take a set of long reads and produce a smaller, better subset. It uses both read length (longer is better) and read identity (higher is better) when choosing which reads pass the filter. https://github.com/rrwick/Filtlong

**Ratatosk** is a de novo error correction tool for erroneous long reads designed for accurate variant calling and assembly. It is based on a compacted and colored de Bruijn graph built from accurate short reads. https://github.com/DecodeGenetics/Ratatosk

**Flye** is a de novo assembler for single-molecule sequencing reads, such as those produced by PacBio and Oxford Nanopore Technologies. It is designed for a wide range of datasets, from small bacterial projects to large mammalian-scale assemblies. https://github.com/mikolmogorov/Flye

**Racon** is as a standalone consensus module to correct raw contigs generated by rapid assembly methods which do not include a consensus step. The goal is to generate genomic consensus which is of similar or better quality compared to the output generated by assembly methods which employ both error correction and consensus steps, while providing a speedup of several times compared to those methods.

**Pilon** is an automated genome assembly improvement and variant detection tool. It significantly improves draft genome assemblies by correcting bases, fixing mis-assemblies and filling gaps. https://github.com/broadinstitute/pilon

**Purge Dups** is a package used to purge haplotigs and overlaps in an assembly based on read depth. https://github.com/dfguan/purge_dups

**RagTag** is a collection of software tools for scaffolding and improving modern genome assemblies. https://github.com/malonge/RagTag

**TGS-GapCloser** is a gap-closing software tool that uses error-prone long reads generated by third-generation-sequence techniques (Pacbio, Oxford Nanopore, etc.) or preassembled contigs to fill N-gap in the genome assembly. https://github.com/BGI-Qingdao/TGS-GapCloser

## Assembly evaluation

**QUAST** (QUality ASsessment Tool) is a widely used software for evaluating and comparing genome  <br>
  &emsp; &ensp;assemblies in bioinformatics, providing metrics on assembly quality, such as contiguity and completeness.  <br>
  &emsp; &ensp;The tool can be used with or without a reference genome and generates detailed reports, tables, and plots  <br>
  &emsp; &ensp;for analysis https://github.com/ablab/quast.

**Merqury** is a tool for validating genomes assembled using long read sequencing. The algorithm uses k-mer<br> &emsp; &ensp;to evaluate base accuracy and completeness of a genome by comparing the de novo assembled genome <br> &emsp; &ensp;with high accuracy reads that are not used in the assembly. It does not require another reference assembly <br> &emsp; &ensp;for its evaluation. It is able to evaluate the haplotype-specific accuracy,completeness, phase block  <br>
  &emsp; &ensp;continuity and switch errors in trio binned assemblies https://github.com/marbl/merqury.
  
**BUSCO**
 (Benchmarking Universal Single-Copy Orthologs) provides measures for quantitative assessment of genome assembly, gene set, and transcriptome completeness based on evolutionarily informed expectations of gene content from near-universal single-copy orthologs, please visit https://busco.ezlab.org/.

**MultiQC** is used create a single report with interactive plots for multiple bioinformatics analyses across many samples. It reports can describe multiple analysis steps and large numbers of samples within a single plot, and multiple analysis tools making it ideal for routine fast quality control. For more information, please visit the MultiQC website https://github.com/MultiQC/MultiQC.
